examples/oauth.rs

```diff
@@ -0,0 +1,261 @@
+
+#[tokio::main]
+async fn main() {
+    // Set the RUST_LOG, if it hasn't been explicitly defined
+    if std::env::var("RUST_LOG").is_err() {
+        std::env::set_var("RUST_LOG", "hello_world=debug")
+    }
+    tracing_subscriber::fmt::init();
+
+    // `MemoryStore` just used as an example. Don't use this in production.
+    let store = MemoryStore::new();
+    let oauth_client = oauth_client();
+    let app = route("/", get(index))
+        .route("/auth/discord", get(discord_auth))
+        .route("/auth/authorized", get(login_authorized))
+        .route("/protected", get(protected))
+        .route("/logout", get(logout))
+        .layer(AddExtensionLayer::new(store))
+        .layer(AddExtensionLayer::new(oauth_client));
+
+    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
+    tracing::debug!("listening on {}", addr);
+
+    axum::Server::bind(&addr)
+        .serve(app.into_make_service())
+        .await
+        .unwrap();
+}
+
+fn oauth_client() -> BasicClient {
+    // Environment variables (* = required):
+    // *"CLIENT_ID"     "123456789123456789";
+    // *"CLIENT_SECRET" "rAn60Mch4ra-CTErsSf-r04utHcLienT";
+    //  "REDIRECT_URL"  "http://127.0.0.1:3000/auth/authorized";
+    //  "AUTH_URL"      "https://discord.com/api/oauth2/authorize?response_type=code";
+    //  "TOKEN_URL"     "https://discord.com/api/oauth2/token";
+
+    let client_id = env::var("CLIENT_ID").expect("Missing CLIENT_ID!");
+    let client_secret = env::var("CLIENT_SECRET").expect("Missing CLIENT_SECRET!");
+    let redirect_url = env::var("REDIRECT_URL")
+        .unwrap_or_else(|_| "http://127.0.0.1:3000/auth/authorized".to_string());
+
+    let auth_url = env::var("AUTH_URL").unwrap_or_else(|_| {
+        "https://discord.com/api/oauth2/authorize?response_type=code".to_string()
+    });
+
+    let token_url = env::var("TOKEN_URL")
+        .unwrap_or_else(|_| "https://discord.com/api/oauth2/token".to_string());
+
+    BasicClient::new(
+        ClientId::new(client_id),
+        Some(ClientSecret::new(client_secret)),
+        AuthUrl::new(auth_url).unwrap(),
+        Some(TokenUrl::new(token_url).unwrap()),
+    )
+    .set_redirect_uri(RedirectUrl::new(redirect_url).unwrap())
+}
+
+// The user data we'll get back from Discord.
+// https://discord.com/developers/docs/resources/user#user-object-user-structure
+#[derive(Debug, Serialize, Deserialize)]
+pub struct User {
+    pub id: String,
+    pub avatar: Option<String>,
+    pub username: String,
+    pub discriminator: String,
+}
+
+// Session is optional
+async fn index(user: Option<User>) -> impl IntoResponse {
+    match user {
+        Some(u) => format!(
+            "Hey {}! You're logged in!\nYou may now access `/protected`.\nLog out with `/logout`.",
+            u.username
+        ),
+        None => "You're not logged in.\nVisit `/auth/discord` to do so.".to_string(),
+    }
+}
+
+pub(crate) async fn discord_auth(client: Extension<BasicClient>) -> impl IntoResponse {
```

Comment:
```suggestion
async fn discord_auth(client: Extension<BasicClient>) -> impl IntoResponse {
```

I don't all these functions need to be `pub(crate)`. Could confuse people.